File: src/extensions/inputs.ts
import type { CommandContext } from '../commands/index.js';
import { loadJson } from '../commands/shared/json.js';
import { readTextFromFileOrStdin } from '../commands/shared/io.js';
import { err } from '../types.js';
import { errorMessage } from '../utils.js';
import { vfsError } from '../vfs/errors.js';
import { formatVfsError } from './errors.js';
import { materializedLimitError } from '../commands/shared/io.js';
import { helperFailure, helperSuccess, type HelperResult } from './types.js';

export async function readTextInput(
  ctx: CommandContext,
  commandName: string,
  filePath: string | undefined,
  stdin: Uint8Array,
): Promise<HelperResult<string>> {
  return loadInput(ctx, commandName, filePath, async function (): Promise<HelperResult<string>> {
    const { text, error } = await readTextFromFileOrStdin(ctx, filePath, stdin, commandName);
    if (error !== undefined) {
      return helperFailure(error);
    }
    return helperSuccess(text ?? '');
  });
}

export async function readBytesInput(
  ctx: CommandContext,
  commandName: string,
  filePath: string | undefined,
  stdin: Uint8Array,
): Promise<HelperResult<Uint8Array>> {
  if (filePath !== undefined) {
    return loadInput(ctx, commandName, filePath, async function (): Promise<HelperResult<Uint8Array>> {
      const info = await ctx.vfs.stat(filePath);
      if (!info.exists) {
        return helperFailure(
          await formatVfsError(ctx, commandName, filePath, vfsError('ENOENT', ctx.vfs.normalize(filePath)), {
            notFoundLabel: 'file not found',
          }),
        );
      }
      if (info.isDir) {
        return helperFailure(
          await formatVfsError(ctx, commandName, filePath, vfsError('EISDIR', ctx.vfs.normalize(filePath))),
        );
      }
      const limitError = materializedLimitError(ctx, commandName, ctx.vfs.normalize(filePath), info.size);
      if (limitError !== undefined) {
        return helperFailure(limitError);
      }
      return helperSuccess(await ctx.vfs.readBytes(filePath));
    });
  }

  if (stdin.length > 0) {
    const limitError = materializedLimitError(ctx, commandName, 'stdin', stdin.length);
    if (limitError !== undefined) {
      return helperFailure(limitError);
    }
    return helperSuccess(new Uint8Array(stdin));
  }

  return helperFailure(err(`${commandName}: no input provided. Usage: see 'help ${commandName}'`));
}

export async function readJsonInput<T = unknown>(
  ctx: CommandContext,
  commandName: string,
  filePath: string | undefined,
  stdin: Uint8Array,
): Promise<HelperResult<T>> {
  return loadInput(ctx, commandName, filePath, async function (): Promise<HelperResult<T>> {
    const { value, error } = await loadJson(ctx, filePath, stdin, commandName);
    if (error !== undefined) {
      return helperFailure(error);
    }
    return helperSuccess(value as T);
  });
}

async function loadInput<T>(
  ctx: CommandContext,
  commandName: string,
  filePath: string | undefined,
  loader: () => Promise<HelperResult<T>>,
): Promise<HelperResult<T>> {
  try {
    return await loader();
  } catch (caught) {
    return handleInputFailure(ctx, commandName, filePath, caught);
  }
}

async function handleInputFailure<T>(
  ctx: CommandContext,
  commandName: string,
  filePath: string | undefined,
  caught: unknown,
): Promise<HelperResult<T>> {
  if (filePath !== undefined) {
    return helperFailure(
      await formatVfsError(ctx, commandName, filePath, caught, {
        notFoundLabel: 'file not found',
      }),
    );
  }

  return helperFailure(err(`${commandName}: ${errorMessage(caught)}`));
}

File: python/src/onetool/extensions/inputs.py
from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import TypeVar, cast

from ..commands.core import CommandContext
from ..commands.shared.io import materialized_limit_error, read_text_from_file_or_stdin
from ..commands.shared.json import load_json
from ..types import err
from ..utils import error_message
from ..vfs.errors import vfs_error
from .errors import FormatVfsErrorOptions, format_vfs_error
from .types import HelperResult, helper_failure, helper_success

T = TypeVar("T")


async def read_text_input(
    ctx: CommandContext,
    command_name: str,
    file_path: str | None,
    stdin: bytes,
) -> HelperResult[str]:
    return await _load_input(
        ctx,
        command_name,
        file_path,
        lambda: _read_text_input_inner(ctx, command_name, file_path, stdin),
    )


async def read_bytes_input(
    ctx: CommandContext,
    command_name: str,
    file_path: str | None,
    stdin: bytes,
) -> HelperResult[bytes]:
    if file_path is not None:
        return await _load_input(
            ctx,
            command_name,
            file_path,
            lambda: _read_file_bytes(ctx, command_name, file_path),
        )

    if stdin:
        limit_error = materialized_limit_error(ctx, command_name, "stdin", len(stdin))
        if limit_error is not None:
            return helper_failure(limit_error)
        return helper_success(bytes(stdin))

    return helper_failure(err(f"{command_name}: no input provided. Usage: see 'help {command_name}'"))


async def read_json_input(
    ctx: CommandContext,
    command_name: str,
    file_path: str | None,
    stdin: bytes,
) -> HelperResult[T]:
    return await _load_input(
        ctx,
        command_name,
        file_path,
        lambda: _read_json_input_inner(ctx, command_name, file_path, stdin),
    )


async def _read_text_input_inner(
    ctx: CommandContext,
    command_name: str,
    file_path: str | None,
    stdin: bytes,
) -> HelperResult[str]:
    loaded = await read_text_from_file_or_stdin(ctx, file_path, stdin, command_name)
    if loaded.error is not None:
        return helper_failure(loaded.error)
    return helper_success(loaded.text or "")


async def _read_file_bytes(
    ctx: CommandContext,
    command_name: str,
    file_path: str,
) -> HelperResult[bytes]:
    info = await ctx.vfs.stat(file_path)
    normalized_path = ctx.vfs.normalize(file_path)

    if not info.exists:
        return helper_failure(
            await format_vfs_error(
                ctx,
                command_name,
                file_path,
                vfs_error("ENOENT", normalized_path),
                FormatVfsErrorOptions(not_found_label="file not found"),
            )
        )

    if info.is_dir:
        return helper_failure(
            await format_vfs_error(
                ctx,
                command_name,
                file_path,
                vfs_error("EISDIR", normalized_path),
            )
        )

    limit_error = materialized_limit_error(ctx, command_name, normalized_path, info.size)
    if limit_error is not None:
        return helper_failure(limit_error)

    return helper_success(await ctx.vfs.read_bytes(file_path))


async def _read_json_input_inner(
    ctx: CommandContext,
    command_name: str,
    file_path: str | None,
    stdin: bytes,
) -> HelperResult[T]:
    loaded = await load_json(ctx, file_path, stdin, command_name)
    if loaded.error is not None:
        return helper_failure(loaded.error)
    return helper_success(cast(T, loaded.value))


async def _load_input(
    ctx: CommandContext,
    command_name: str,
    file_path: str | None,
    loader: Callable[[], Awaitable[HelperResult[T]]],
) -> HelperResult[T]:
    try:
        return await loader()
    except Exception as caught:
        return await _handle_input_failure(ctx, command_name, file_path, caught)


async def _handle_input_failure(
    ctx: CommandContext,
    command_name: str,
    file_path: str | None,
    caught: object,
) -> HelperResult[T]:
    if file_path is not None:
        return helper_failure(
            await format_vfs_error(
                ctx,
                command_name,
                file_path,
                caught,
                FormatVfsErrorOptions(not_found_label="file not found"),
            )
        )

    return helper_failure(err(f"{command_name}: {error_message(caught)}"))
